ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/recovery/arjunacore/SubordinationManagerXAResourceOrphanFilter.java

```diff
@@ -0,0 +1,112 @@
+
+import com.arjuna.ats.arjuna.coordinator.TxControl;
+import com.arjuna.ats.arjuna.recovery.RecoveryManager;
+import com.arjuna.ats.arjuna.recovery.RecoveryModule;
+import com.arjuna.ats.internal.jta.transaction.arjunacore.jca.SubordinateTransaction;
+import com.arjuna.ats.internal.jta.transaction.arjunacore.jca.SubordinationManager;
+import com.arjuna.ats.jta.recovery.XAResourceOrphanFilter;
+import com.arjuna.ats.jta.xa.XATxConverter;
+import com.arjuna.ats.jta.xa.XidImple;
+
+import javax.transaction.xa.XAException;
+import javax.transaction.xa.Xid;
+import java.util.Vector;
+
+/**
+ * An XAResourceOrphanFilter which vetos rollback for xids which have an in-flight subordinate transaction.
+ * <p>
+ * The SubordinateAtomicActionRecoveryModule must be loaded and in a position prior to the XARecoveryModule within the list
+ * of recovery modules for this to work so we verify that during orphan detection.
+ */
+public class SubordinationManagerXAResourceOrphanFilter implements XAResourceOrphanFilter {
+    private boolean validated;
+    private SubordinateAtomicActionRecoveryModule subordinateAtomicActionRecoveryModule;
+
+    @Override
+    public Vote checkXid(Xid xid) {
+        if (xid.getFormatId() != XATxConverter.FORMAT_ID) {
+            // we only care about Xids created by the JTA
+            return Vote.ABSTAIN;
+        }
+
+        if (!TxControl.getXANodeName().equals(XATxConverter.getSubordinateNodeName(new XidImple(xid).getXID()))) {
+            // It either doesn't have a subordinate node name or isn't for this server
+            return Vote.ABSTAIN;
+        }
+
+        if (!validatePosition()) {
+            return Vote.LEAVE_ALONE;
+        }
+
+        if (!subordinateAtomicActionRecoveryModule.isRecoverySucceeded()) {
+            return Vote.LEAVE_ALONE;
+        }
+
+        XidImple theXid = new XidImple(xid);
+        SubordinateTransaction importedTransaction = null;
+        try {
+            importedTransaction = SubordinationManager.getTransactionImporter().getImportedTransaction(theXid);
+        } catch (XAException e) {
+            return Vote.LEAVE_ALONE;
+        }
+        if (importedTransaction != null) {
+            return Vote.LEAVE_ALONE;
+        } else {
+            return Vote.ROLLBACK;
+        }
+    }
+
+    /**
+     * It is important to verify that the SubordinateAtomicActionRecoveryModule has been loaded before the XARecoveryModule
+     * as otherwise when we check for an imported transaction it may not have been loaded by the  SubordinateAtomicActionRecoveryModule yet.
+     *
+     * This method also stores a reference to the SubordinateAtomicActionRecoveryModule so that we can verify there
+     * were no failures in recovery transactions for this pass of the recovery cycle.
+     *
+     * @return true if and only if SubordinateAtomicActionRecoveryModule appears before XARecoveryModule in the modules
+     */
+    private boolean validatePosition() {
```

Comment:
I still say it is a complicated bit of code. All it is achieving is:
```java
return modules.contains(SAARM) && modules.indexOf(SAARM) < modules.indexOf(XARM))
```

Reply:
I don't think I can do that because I don't have a reference for SAARM and XARM (I am having to compare the classnames)

Reply:
I wasn't suggesting that my line was an alternative (just that the routine is achieving something quite easy with lots of lines of code and therefore harder to maintain and read. My original comment said:

> Can we either refactor the code or simplify.

Reply:
> can we refactor or simplify
maybe I can move this to the SubordinateAtomicActionRecoveryModule but the code will be the same more or less in that location, plus I will still need to scan the list to obtain the instance of SubordinateAtomicActionRecoveryModule to check for errors.

We do already do similar code elsewhere:
https://github.com/jbosstm/narayana/blob/master/ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/resources/arjunacore/XAResourceRecord.java#L1254

Reply:
> Blaargh! There are better ways to do this!

Someone had similar thought to me here. Anyway, I don't want to labour the issue so I am approving.